### ISSUE TITLE
journald should distinguish service log msgs written to stdout and stderr

### ISSUE THREAD

**Issue by narun4sk** [systemd/systemd#5019]
### Submission type

  - [ ] Bug report
  - [x] Request for enhancement (RFE)

### systemd version the issue has been seen with

```
# systemd --version
systemd 231
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN
```

### Used distribution

```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux testing (stretch)
Release:	testing
Codename:	stretch
```

## Feature Request: please add _TRANSPORT=stderr

Please assume there's an app/script which prints a great deal of logging data to `stdout`. When something bad but expected happens within the script (exception handling), script reports an error to `stderr` and then continues doing what it was doing.

`systemd` for sure is capable to collect this data and put it into the `journal`:

    # systemctl show sd_test.service | grep 'Standard[OE]'
    StandardOutput=journal
    StandardError=inherit

My `sd_test.service`:

    [Unit]
    Description=A Test Service simply printing to stdout and stderr
    
    [Service]
    Type=simple
    ExecStart=/home/narunas/sd_test.py
    
    [Install]
    WantedBy=multi-user.target

`journalctl` now has required data:

    # systemctl reenable sd_test.service
    # systemctl restart sd_test.service
    # journalctl -n -u sd_test
    -- Logs begin at Fri 2016-11-11 15:49:33 GMT, end at Tue 2017-01-03 19:23:18 GMT. --
    Jan 03 19:23:01 dev-box sd_test.py[13183]: This is "stderr": 2
    Jan 03 19:23:02 dev-box sd_test.py[13183]: This is "stdout": 2
    Jan 03 19:23:03 dev-box sd_test.py[13183]: This is "stderr": 3
    Jan 03 19:23:07 dev-box sd_test.py[13183]: This is "stdout": 3
    Jan 03 19:23:07 dev-box sd_test.py[13183]: This is "stderr": 4
    Jan 03 19:23:11 dev-box sd_test.py[13183]: This is "stdout": 4
    Jan 03 19:23:12 dev-box sd_test.py[13183]: This is "stderr": 5
    Jan 03 19:23:15 dev-box sd_test.py[13183]: This is "stdout": 5
    Jan 03 19:23:17 dev-box sd_test.py[13183]: This is "stdout": 6
    Jan 03 19:23:18 dev-box sd_test.py[13183]: This is "stderr": 6

I know there's `_TRANSPORT` journal field, sadly `stderr` is tagged with the same tag:

> **stdout**   
    for those read from a service's standard output or error output

I can of course change my `ExecStart` like so:

    ExecStart=/bin/bash -c '/home/narunas/sd_test.py 2> /some/file/path'

But that is not ideal as I'm mixing up `journal` with files...

With `-o verbose` option:

    Tue 2017-01-03 20:23:49.994171 GMT [s=0b15e5c69e2f476eb200d2bdda769465;i=12e28;b=db55b41f61144ae69cc86acfb75209fb;m=42e034d52dc;t=5453672322abb;x=8b696c5447bc2bce]
        PRIORITY=6
        _UID=0
        _GID=0
        _CAP_EFFECTIVE=3fffffffff
        _SYSTEMD_SLICE=system.slice
        _BOOT_ID=db55b41f61144ae69cc86acfb75209fb
        _MACHINE_ID=c5a9e78e2c854065a9b041c58f07c2b2
        _HOSTNAME=dev-box
        SYSLOG_FACILITY=3
        _TRANSPORT=stdout
        SYSLOG_IDENTIFIER=sd_test.py
        _COMM=python3
        _EXE=/usr/bin/python3.5
        _CMDLINE=python3 /home/narunas/sd_test.py
        _SYSTEMD_CGROUP=/system.slice/sd_test.service
        _SYSTEMD_UNIT=sd_test.service
        MESSAGE=This is "stdout": 729
        _PID=13183
    Tue 2017-01-03 20:23:52.979981 GMT [s=0b15e5c69e2f476eb200d2bdda769465;i=12e29;b=db55b41f61144ae69cc86acfb75209fb;m=42e037ae22e;t=54536725fba0d;x=8802c0df56848907]
        PRIORITY=6
        _UID=0
        _GID=0
        _CAP_EFFECTIVE=3fffffffff
        _SYSTEMD_SLICE=system.slice
        _BOOT_ID=db55b41f61144ae69cc86acfb75209fb
        _MACHINE_ID=c5a9e78e2c854065a9b041c58f07c2b2
        _HOSTNAME=dev-box
        SYSLOG_FACILITY=3
        _TRANSPORT=stdout
        SYSLOG_IDENTIFIER=sd_test.py
        _COMM=python3
        _EXE=/usr/bin/python3.5
        _CMDLINE=python3 /home/narunas/sd_test.py
        _SYSTEMD_CGROUP=/system.slice/sd_test.service
        _SYSTEMD_UNIT=sd_test.service
        MESSAGE=This is "stderr": 745
        _PID=13183

Note that apart my script's silly *MESSAGE* there's nothing useful that would let to distinguish what is `stderr` or `stdout`.

Could you recommend a better way to view `stderr` log only? Is it a bad practice to seek `stderr` logs?

*EDIT*:

Please also see the following [thread](http://unix.stackexchange.com/questions/334621/how-to-view-stderr-log-only-with-journalctl).

**Comment by poettering** [comment]
So you want a way to distinguish messages written to stdout and stderr, did I get this right?

**Comment by narun4sk** [comment]
Yes, absolutely correct, a way to distinguish messages written to stdout and stderr, thank you.

**Comment by 2fours** [comment]
+1

**Comment by HenrikHolst** [comment]
I think a good idea would be to add output from stderr with a level equivalent of LOG_WARNING or LOG_ERR like the journal does with logs from syslog at the moment.

**Comment by HenrikHolst** [comment]
Looking into this a little more and it turns out that the journal have this kind of support for "new-style" daemons where the output from say stderr can be prefixed with a <level> (see man 3 sd-daemon) so question is if the journal should threat output from stderr without such prefixes as SD_WARNING instead of defaulting to SD_INFO so that there is a distinction even of the "old-style" daemons (after all logging to stderr vs stdout should be for warnings/errors).

**Comment by boucman** [comment]
most daemon use either stdout or stderr, afaict, so doing that would mean flooding the log with warning level messages, since they don't use the stdout vs stderr to reflect the importance of the message...

**Comment by waqarkhan3389** [comment]
+1

**Comment by pabl0** [comment]
One feature I would love to see: an option (e.g. --stderr-priority=) for systemd-cat which would set the priority for input from stderr. Perhaps the default priority could be "err". That way you could pass the output of a script ((a backup cron job, for instance) directly to the journal and have any errors or warnings stand out.

**Comment by florianjacob** [comment]
At least in my case, it would suffice to be able to specify `SyslogLevel=` separately for for `StandardOutput` and `StandardError`, so I could leave `StandardOutput` at priority `INFO` (6) and make `StandardError` priority `NOTICE` (5).

**Comment by mgsloan** [comment]
@pabl0 FYI I've opened a PR adding the `--stderr-priority` option that you suggested - https://github.com/systemd/systemd/pull/11336 . I don't think it really properly addresses the overall issue here, but it might help a bit.

**Comment by florianjacob** [comment]
For the sake of back-referencing, poettering explained an idea to fully address this issue at https://github.com/systemd/systemd/pull/6599#issuecomment-438376141 .

**Comment by jamestiotio** [comment]
I might also take a look at this whenever I have the time to do so. Will take into account Poettering's suggestion to preserve the "global ordering" property.

**Comment by piller-imre** [comment]
I also would like to find an elegant solution to this problem. In my opinion, an optional settings for the default priority of the stderr outputs could be a viable solution, like @florianjacob has mentioned.

As a workaround, I have solved the problem by using a simple shell script, which is practically a wrapper around the original service. Let assume that, the name of the service is `service`. By using the following redirection the stdout and stderr have been distinguished by priority levels:
```bash
service 2| sed 's/^/<3>/g'
```
(It assumes that the stderr does not use explicit priority level prefixes.)
